File: src/main/java/space/rogi27/homabric/commands/BaseCommands.java
package space.rogi27.homabric.commands;

import com.mojang.brigadier.arguments.StringArgumentType;
import com.mojang.brigadier.builder.LiteralArgumentBuilder;
import com.mojang.brigadier.context.CommandContext;
import com.mojang.brigadier.exceptions.CommandSyntaxException;
import com.mojang.brigadier.suggestion.Suggestions;
import com.mojang.brigadier.suggestion.SuggestionsBuilder;
import eu.pb4.sgui.api.gui.SimpleGui;
import me.lucko.fabric.api.permissions.v0.Permissions;
import net.fabricmc.fabric.api.command.v1.CommandRegistrationCallback;
import net.minecraft.command.EntitySelector;
import net.minecraft.command.argument.EntityArgumentType;
import net.minecraft.command.argument.IdentifierArgumentType;
import net.minecraft.server.command.ServerCommandSource;
import net.minecraft.server.network.ServerPlayerEntity;
import net.minecraft.text.LiteralText;
import net.minecraft.text.TranslatableText;
import net.minecraft.util.Formatting;
import net.minecraft.util.Identifier;
import net.minecraft.util.registry.Registry;
import space.rogi27.homabric.Homabric;
import space.rogi27.homabric.config.HomabricConfig;
import space.rogi27.homabric.utils.HomeObject;
import space.rogi27.homabric.utils.PlayerObject;

import java.util.ArrayList;
import java.util.concurrent.CompletableFuture;

import static net.minecraft.server.command.CommandManager.argument;
import static net.minecraft.server.command.CommandManager.literal;

public class BaseCommands {
    public static void init() {
        CommandRegistrationCallback.EVENT.register((dispatcher, dedicated) -> {
            dispatcher.register(bindHomeToolsTo("home"));
            dispatcher.register(bindHomeToolsTo("h"));
        });
    }

    public static LiteralArgumentBuilder<ServerCommandSource> bindHomeToolsTo(String name) {
        LiteralArgumentBuilder<ServerCommandSource> item = literal(name)
                .requires(Permissions.require("homabric.base.use", true))
                .then(literal("set")
                        .requires(Permissions.require("homabric.base.set", true))
                        .then(argument("home", StringArgumentType.word())
                                .suggests(BaseCommands::suggestPlayerHomes)
                                .executes((BaseCommands::set)))
                        .executes(BaseCommands::set))
                .then(literal("remove")
                        .requires(Permissions.require("homabric.base.remove", true))
                        .then(argument("home", StringArgumentType.word())
                                .suggests(BaseCommands::suggestPlayerHomes)
                                .executes((BaseCommands::remove))
                        )
                )
                .then(literal("list")
                        .requires(Permissions.require("homabric.base.list", true))
                        .executes((BaseCommands::list))
                )
                .then(literal("allow")
                        .requires(Permissions.require("homabric.base.allow", true))
                        .then(argument("player", EntityArgumentType.player())
                            .suggests(BaseCommands::suggestOnlinePlayerStrings)
                            .then(argument("home", StringArgumentType.word())
                                    .suggests(BaseCommands::suggestPlayerHomes)
                                    .executes((BaseCommands::allowHome))
                            )
                        )
                )
                .then(literal("disallow")
                        .requires(Permissions.require("homabric.base.disallow", true))
                        .then(argument("player", StringArgumentType.word())
                                .suggests(BaseCommands::suggestOnlinePlayerStrings)
                                .then(argument("home", StringArgumentType.word())
                                        .suggests(BaseCommands::suggestPlayerHomes)
                                        .executes((BaseCommands::disallowHome))
                                )
                        )
                )
                .then(literal("p")
                        .requires(Permissions.require("homabric.base.others", true))
                        .then(argument("player", StringArgumentType.word())
                            .suggests(BaseCommands::suggestOnlinePlayerStrings)
                            .then(argument("home", StringArgumentType.word())
                                        .suggests(BaseCommands::suggestAllowedHomes)
                                        .executes((BaseCommands::teleportToAllowed))
                            )
                        )
                )
                .then(literal("setIcon")
                        .requires(Permissions.require("homabric.base.setIcon", true))
                        .then(argument("home", StringArgumentType.word())
                                .suggests(BaseCommands::suggestPlayerHomes)
                                .then(argument("item", IdentifierArgumentType.identifier())
                                        .executes((BaseCommands::setIcon))
                                )
                        )
                )
                .then(argument("home", StringArgumentType.word())
                        .suggests(BaseCommands::suggestPlayerHomes)
                        .requires(Permissions.require("homabric.base.byName", true))
                        .executes((BaseCommands::teleportByName))
                )
                .executes((BaseCommands::teleport));
        return item;
    }

    public static CompletableFuture<Suggestions> suggestOnlinePlayerStrings(CommandContext<ServerCommandSource> context, SuggestionsBuilder builder) {
        context.getSource().getServer().getPlayerManager().getPlayerList().forEach((player) -> {
            builder.suggest(player.getEntityName());
        });
        return builder.buildFuture();
    }


    public static CompletableFuture<Suggestions> suggestPlayerHomes(CommandContext<ServerCommandSource> context, SuggestionsBuilder builder) {
        PlayerObject player = HomabricConfig.Config.getPlayer(context.getSource().getName());
        if(player != null) player.getHomeNames().forEach(builder::suggest);
        return builder.buildFuture();
    }

    public static CompletableFuture<Suggestions> suggestAllowedHomes(CommandContext<ServerCommandSource> context, SuggestionsBuilder builder) throws CommandSyntaxException {
        String owner = context.getArgument("player", String.class);
        if(owner != null && HomabricConfig.Config.getPlayer(owner) != null) {
            ArrayList<String> allowedHomes = HomabricConfig.Config
                    .getPlayer(owner)
                    .getAllowedHomeNames(context.getSource().getEntity().getEntityName());

            allowedHomes.forEach(builder::suggest);
        }
        return builder.buildFuture();
    }

    public static int setIcon(CommandContext<ServerCommandSource> context) throws CommandSyntaxException {
        String homeName = context.getArgument("home", String.class);
        HomeObject home = HomabricConfig.Config.getOrCreatePlayer(context.getSource().getName()).getHome(homeName);

        if(home == null) {
            context.getSource().sendFeedback(new TranslatableText("text.homabric.no_home"), false);
            return 1;
        }

        HomeObject.IconResult result = home.setIcon(context.getArgument("item", Identifier.class));
        switch (result) {
            case WRONG_ICON -> {
                context.getSource().sendFeedback(new TranslatableText("text.homabric.no_icon"), false);
            }
            case ICON_SET -> {
                context.getSource().sendFeedback(new TranslatableText("text.homabric.icon_changed", new LiteralText(homeName).formatted(Formatting.WHITE), Registry.ITEM.get(context.getArgument("item", Identifier.class)).getName().shallowCopy().formatted(Formatting.AQUA)).formatted(Formatting.GREEN), false);
            }
        }

        Homabric.reloadConfig();
        return 1;
    }

    public static int teleport(CommandContext<ServerCommandSource> context) throws CommandSyntaxException {
        String homeName = "home";
        PlayerObject player = HomabricConfig.Config.getOrCreatePlayer(context.getSource().getName());
        PlayerObject.TeleportResult result = player.teleportToHome(context.getSource(), homeName);

        switch (result) {
            case NO_HOME -> {
                context.getSource().sendFeedback(new TranslatableText("text.homabric.no_home"), false);
            }
            case TELEPORT_DONE -> {
                context.getSource().sendFeedback(new TranslatableText("text.homabric.teleport_done", new LiteralText(homeName).formatted(Formatting.WHITE)).formatted(Formatting.GREEN), false);
            }
        }
        return 1;
    }

    public static int teleportByName(CommandContext<ServerCommandSource> context) throws CommandSyntaxException {
            String homeName = context.getArgument("home", String.class);
            PlayerObject player = HomabricConfig.Config.getOrCreatePlayer(context.getSource().getName());
            PlayerObject.TeleportResult result = player.teleportToHome(context.getSource(), homeName);

            switch (result) {
                case NO_HOME -> {
                    context.getSource().sendFeedback(new TranslatableText("text.homabric.no_home"), false);
                }
                case TELEPORT_DONE -> {
                    context.getSource().sendFeedback(new TranslatableText("text.homabric.teleport_done", new LiteralText(homeName).formatted(Formatting.WHITE)).formatted(Formatting.GREEN), false);
                }
            }
            return 1;
    }

    public static int teleportToAllowed(CommandContext<ServerCommandSource> context) throws CommandSyntaxException {
        String owner = context.getArgument("player", String.class);
        String homeName = context.getArgument("home", String.class);

        PlayerObject.TeleportToOtherResult result = PlayerObject.teleportToOtherHome(context.getSource(), owner, homeName, false);

        switch (result) {
            case NO_PLAYER -> {
                context.getSource().sendFeedback(new TranslatableText("text.homabric.no_player_exists").formatted(Formatting.RED), false);
            }
            case NO_HOME -> {
                context.getSource().sendFeedback(new TranslatableText("text.homabric.no_home"), false);
            }
            case NO_ACCESS -> {
                context.getSource().sendFeedback(new TranslatableText("text.homabric.no_home_access").formatted(Formatting.RED), false);
            }
            case TELEPORT_DONE -> {
                context.getSource().sendFeedback(new TranslatableText("text.homabric.teleport_done", new LiteralText(homeName).formatted(Formatting.WHITE)).formatted(Formatting.GREEN), false);
            }
        }
        return 1;
    }

    public static int set(CommandContext<ServerCommandSource> context) throws CommandSyntaxException {
        if(!context.getSource().getEntity().isPlayer()) {
            return 0;
        }

        String homeName = "home";
        try {
            if(context.getArgument("home", String.class) != null) homeName = context.getArgument("home", String.class);
        } catch (Exception ex) {
            Homabric.LOGGER.warn("Using command without name, referencing home");
        }

        PlayerObject player = HomabricConfig.Config.getOrCreatePlayer(context.getSource().getName());
        if(player.isLimitReached(context.getSource())) {
            context.getSource().sendFeedback(new TranslatableText("text.homabric.home_limit_reached").formatted(Formatting.RED), false);
            return 1;
        }

        PlayerObject.HomeCreationResult result = player.createOrUpdateHome(context.getSource(), homeName);

        if(result == PlayerObject.HomeCreationResult.HOME_CREATED) {
            context.getSource().sendFeedback(new TranslatableText("text.homabric.new_home_created", new LiteralText(homeName).formatted(Formatting.WHITE)).formatted(Formatting.GREEN), false);
        } else {
            context.getSource().sendFeedback(new TranslatableText("text.homabric.home_location_updated", new LiteralText(homeName).formatted(Formatting.WHITE)).formatted(Formatting.GREEN), false);
        }

        Homabric.reloadConfig();
        return 1;
    }

    public static int remove(CommandContext<ServerCommandSource> context) throws CommandSyntaxException {
        if(!context.getSource().getEntity().isPlayer()) {
            return 0;
        }

        String homeName = context.getArgument("home", String.class);
        PlayerObject.HomeRemoveResult result = HomabricConfig.Config.getOrCreatePlayer(context.getSource().getName()).removeHome(homeName);

        switch (result) {
            case NO_HOME -> {
                context.getSource().sendFeedback(new TranslatableText("text.homabric.no_home").formatted(Formatting.RED), false);
            }
            case HOME_REMOVED -> {
                context.getSource().sendFeedback(new TranslatableText("text.homabric.home_removed", new LiteralText(homeName).formatted(Formatting.WHITE)).formatted(Formatting.GREEN), false);
            }
        }
        return 1;
    }

    public static int list(CommandContext<ServerCommandSource> context) throws CommandSyntaxException {
        PlayerObject playerData = HomabricConfig.Config.getPlayer(context.getSource().getName());
        if (playerData == null) {
            context.getSource().sendFeedback(new TranslatableText("text.homabric.no_homes").formatted(Formatting.YELLOW), false);
            return 1;
        }

        SimpleGui gui = playerData.getHomesGUI(context.getSource());
        gui.open();
        return 1;
    }

    public static int allowHome(CommandContext<ServerCommandSource> context) throws CommandSyntaxException {
        String homeName = context.getArgument("home", String.class);
        ServerPlayerEntity allowedPlayer = context.getArgument("player", EntitySelector.class).getPlayer(context.getSource());

        PlayerObject.HomeAllowResult result = HomabricConfig.Config.getOrCreatePlayer(context.getSource().getName()).allowHome(homeName, allowedPlayer);

        switch (result) {
            case NO_PLAYER -> {
                context.getSource().sendFeedback(new TranslatableText("text.homabric.no_player_exists").formatted(Formatting.RED), false);
            }
            case NO_SELF_ALLOW -> {
                context.getSource().sendFeedback(new TranslatableText("text.homabric.allow_self").formatted(Formatting.BLUE), false);
            }
            case NO_HOME -> {
                context.getSource().sendFeedback(new TranslatableText("text.homabric.no_home").formatted(Formatting.RED), false);
            }
            case ALREADY_ALLOWED -> {
                context.getSource().sendFeedback(new TranslatableText("text.homabric.already_allowed").formatted(Formatting.RED), false);
            }
            case HOME_ALLOWED -> {
                context.getSource().sendFeedback(new TranslatableText("text.homabric.allowed", new LiteralText(homeName).formatted(Formatting.WHITE), new LiteralText(allowedPlayer.getEntityName()).formatted(Formatting.AQUA)).formatted(Formatting.GREEN), false);
            }
        }

        return 1;
    }

    public static int disallowHome(CommandContext<ServerCommandSource> context) throws CommandSyntaxException {
        String homeName = context.getArgument("home", String.class);
        String disallowedPlayer = context.getArgument("player", String.class);

        PlayerObject.HomeDisallowResult result = HomabricConfig.Config.getOrCreatePlayer(context.getSource().getName()).disallowHome(homeName, disallowedPlayer);

        switch (result) {
            case NO_PLAYER -> {
                context.getSource().sendFeedback(new TranslatableText("text.homabric.no_player_exists").formatted(Formatting.RED), false);
            }
            case NO_HOME -> {
                context.getSource().sendFeedback(new TranslatableText("text.homabric.no_home").formatted(Formatting.RED), false);
            }
            case NOT_ALLOWED -> {
                context.getSource().sendFeedback(new TranslatableText("text.homabric.no_player_disallow").formatted(Formatting.RED), false);
            }
            case HOME_ALLOWED -> {
                context.getSource().sendFeedback(new TranslatableText("text.homabric.disallowed", new LiteralText(homeName).formatted(Formatting.WHITE), new LiteralText(disallowedPlayer).formatted(Formatting.AQUA)).formatted(Formatting.GREEN), false);
            }
        }

        return 1;
    }
}

File: src/main/kotlin/space/rogi27/homabric/commands/BaseCommands.kt
package space.rogi27.homabric.commands

import com.mojang.brigadier.Command
import com.mojang.brigadier.arguments.ArgumentType
import com.mojang.brigadier.arguments.StringArgumentType
import com.mojang.brigadier.builder.LiteralArgumentBuilder
import com.mojang.brigadier.context.CommandContext
import com.mojang.brigadier.exceptions.CommandSyntaxException
import com.mojang.brigadier.suggestion.SuggestionsBuilder
import eu.pb4.sgui.api.gui.SimpleGui
import me.lucko.fabric.api.permissions.v0.Permissions
import net.fabricmc.fabric.api.command.v2.CommandRegistrationCallback
import net.minecraft.ChatFormatting
import net.minecraft.commands.CommandSourceStack
import net.minecraft.commands.Commands
import net.minecraft.commands.arguments.EntityArgument
import net.minecraft.commands.arguments.ResourceLocationArgument
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.network.chat.Component
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.item.ItemStack
import space.rogi27.homabric.config.ConfigManager
import space.rogi27.homabric.config.HomabricConfig
import space.rogi27.homabric.config.HomesConfig
import space.rogi27.homabric.helpers.CommandUtils.getHomeArg
import space.rogi27.homabric.helpers.CommandUtils.getPlayerObject
import space.rogi27.homabric.helpers.Completables.suggestAllowedHomes
import space.rogi27.homabric.helpers.Completables.suggestOnlinePlayerStrings
import space.rogi27.homabric.helpers.Completables.suggestPlayerHomes
import space.rogi27.homabric.objects.HomeObject
import space.rogi27.homabric.objects.PlayerObject
import space.rogi27.homabric.polyfills.PermissionLevel

object BaseCommands: RegistrableCommand {
    override fun register(): Boolean {
        CommandRegistrationCallback.EVENT.register(CommandRegistrationCallback { dispatcher, _, _ ->
            HomabricConfig.entries.commandHomeAliases.forEach { alias -> dispatcher.register(registerTeleport(alias)) }
            HomabricConfig.entries.commandHomesAliases.forEach { alias -> dispatcher.register(registerManagement(alias)) }
        })
        return true
    }

    fun registerManagement(name: String): LiteralArgumentBuilder<CommandSourceStack> {
        return Commands.literal(name)
            .requires(Permissions.require("homabric.base.use", PermissionLevel.ALL))
            .executes { context -> list(context) }
            .then(
                Commands.literal("set")
                    .requires(Permissions.require("homabric.base.set", PermissionLevel.ALL))
                    .executes { context -> set(context) }
                    .then(
                        Commands.argument("home", StringArgumentType.word())
                            .suggests { context, suggestionsBuilder -> suggestPlayerHomes(context, suggestionsBuilder) }
                            .executes { context -> set(context) }
                    )
            )
            .then(
                Commands.literal("remove")
                    .requires(Permissions.require("homabric.base.remove", PermissionLevel.ALL))
                    .then(
                        Commands.argument("home", StringArgumentType.word())
                            .suggests { context, suggestionsBuilder -> suggestPlayerHomes(context, suggestionsBuilder) }
                            .executes { context -> remove(context) }
                    )
            )
            .then(
                Commands.literal("allow")
                    .requires(Permissions.require("homabric.base.allow", PermissionLevel.ALL))
                    .then(
                        Commands.argument("player", EntityArgument.player())
                            .suggests { context, suggestionsBuilder -> suggestOnlinePlayerStrings(context, suggestionsBuilder) }
                            .then(
                                Commands.argument("home", StringArgumentType.word())
                                    .suggests { context, suggestionsBuilder -> suggestPlayerHomes(context, suggestionsBuilder) }
                                    .executes { context -> allow(context) }
                            )
                    )
            )
            .then(
                Commands.literal("disallow")
                    .requires(Permissions.require("homabric.base.disallow", PermissionLevel.ALL))
                    .then(
                        Commands.argument("player", StringArgumentType.word())
                            .suggests { context, suggestionsBuilder -> suggestOnlinePlayerStrings(context, suggestionsBuilder) }
                            .then(
                                Commands.argument("home", StringArgumentType.word())
                                    .suggests { context, suggestionsBuilder -> suggestPlayerHomes(context, suggestionsBuilder) }
                                    .executes { context -> disallow(context) }
                            )
                    )
            )
            .then(
                Commands.literal("setIcon")
                    .requires(Permissions.require("homabric.base.setIcon", PermissionLevel.ALL))
                    .then(
                        Commands.argument("home", StringArgumentType.word())
                            .suggests { context, suggestionsBuilder -> suggestPlayerHomes(context, suggestionsBuilder) }
                            .then(
                                Commands.argument("item", ResourceLocationArgument.id())
                                    .executes { context -> setIcon(context) }
                            )
                    )
            )
            .then(
                Commands.literal("p")
                    .requires(Permissions.require("homabric.base.others", PermissionLevel.ALL))
                    .then(
                        Commands.argument("player", StringArgumentType.word())
                            .suggests { context, suggestionsBuilder -> suggestOnlinePlayerStrings(context, suggestionsBuilder) }
                            .then(
                                Commands.argument("home", StringArgumentType.word())
                                    .suggests { context, suggestionsBuilder -> suggestAllowedHomes(context, suggestionsBuilder) }
                                    .executes { context -> teleportToAllowed(context) }
                            )
                    )
            )
    }

    fun registerTeleport(name: String): LiteralArgumentBuilder<CommandSourceStack> {
        return Commands.literal(name)
            .requires(Permissions.require("homabric.base.use", PermissionLevel.ALL))
            .executes { context -> teleport(context, false) }
            .then(
                Commands.argument("home", StringArgumentType.word())
                    .requires(Permissions.require("homabric.base.byName", PermissionLevel.ALL))
                    .suggests { context, suggestionsBuilder -> suggestPlayerHomes(context, suggestionsBuilder) }
                    .executes { context -> teleport(context, true) }
            )
    }

    @Throws(CommandSyntaxException::class)
    fun setIcon(context: CommandContext<CommandSourceStack>): Int {
        val homeName = context.getHomeArg()
        val player = context.getPlayerObject() ?: return 0
        val home: HomeObject? = player.getHome(homeName)
        if (home == null) {
            context.source.sendSystemMessage(Component.translatable("text.homabric.no_home").withStyle(ChatFormatting.RED))
            return 1
        }
        when (home.setIcon(context.getArgument("item", ResourceLocation::class.java))) {
            HomeObject.IconResult.WRONG_ICON -> {
                context.source.sendSystemMessage(Component.translatable("text.homabric.no_icon"))
            }
            HomeObject.IconResult.ICON_SET -> {
                context.source.sendSystemMessage(
                    Component.translatable(
                        "text.homabric.icon_changed", Component.literal(homeName).withStyle(
                            ChatFormatting.WHITE
                        ), ItemStack(BuiltInRegistries.ITEM.get(context.getArgument(
                            "item", ResourceLocation::class.java
                        ))).displayName.copy().withStyle(ChatFormatting.AQUA)
                    ).withStyle(ChatFormatting.GREEN)
                )
            }
        }
        HomesConfig.save()
        return 1
    }

    @Throws(CommandSyntaxException::class)
    fun teleport(context: CommandContext<CommandSourceStack>, hasName: Boolean): Int {
        val player = context.getPlayerObject() ?: return 0
        val homeName = context.getHomeArg()
        val result = player.teleportToHome(context.source, homeName)
        when (result) {
            PlayerObject.TeleportResult.NO_PLAYER -> {
                context.source.sendSystemMessage(
                    Component.translatable("text.homabric.no_player_exists").withStyle(ChatFormatting.RED)
                )
            }
            PlayerObject.TeleportResult.NO_HOME -> {
                context.source.sendSystemMessage(Component.translatable("text.homabric.no_home").withStyle(ChatFormatting.RED))
            }
            PlayerObject.TeleportResult.TELEPORT_DONE -> {}
        }
        return 1
    }

    @Throws(CommandSyntaxException::class)
    fun teleportToAllowed(context: CommandContext<CommandSourceStack>): Int {
        val owner = context.getArgument("player", String::class.java)
        val homeName = context.getHomeArg()
        val result: PlayerObject.TeleportToOtherResult = PlayerObject.teleportToOtherHome(context.source, owner, homeName, false)
        when (result) {
            PlayerObject.TeleportToOtherResult.NO_PLAYER -> {
                context.source.sendSystemMessage(
                    Component.translatable("text.homabric.no_player_exists").withStyle(ChatFormatting.RED)
                )
            }
            PlayerObject.TeleportToOtherResult.NO_HOME -> {
                context.source.sendSystemMessage(Component.translatable("text.homabric.no_home").withStyle(ChatFormatting.RED))
            }
            PlayerObject.TeleportToOtherResult.NO_ACCESS -> {
                context.source.sendSystemMessage(
                    Component.translatable("text.homabric.no_home_access").withStyle(ChatFormatting.RED)
                )
            }
            PlayerObject.TeleportToOtherResult.TELEPORT_DONE -> {}
        }
        return 1
    }

    @Throws(CommandSyntaxException::class)
    fun set(context: CommandContext<CommandSourceStack>): Int {
        val player = context.getPlayerObject() ?: return 0
        val homeName = context.getHomeArg()
        if (player.checkHomeLimit(context.source)) {
            context.source.sendSystemMessage(
                Component.translatable("text.homabric.home_limit_reached").withStyle(ChatFormatting.RED)
            )
            return 1
        }
        val result: PlayerObject.HomeCreationResult = player.createOrUpdateHome(context.source, homeName)
        if (result === PlayerObject.HomeCreationResult.HOME_CREATED) {
            context.source.sendSystemMessage(
                Component.translatable(
                    "text.homabric.new_home_created", Component.literal(homeName).withStyle(
                        ChatFormatting.WHITE
                    )
                ).withStyle(ChatFormatting.GREEN)
            )
        } else {
            context.source.sendSystemMessage(
                Component.translatable(
                    "text.homabric.home_location_updated", Component.literal(homeName).withStyle(
                        ChatFormatting.WHITE
                    )
                ).withStyle(ChatFormatting.GREEN)
            )
        }
        ConfigManager.saveAndLoadAll()
        return 1
    }

    @Throws(CommandSyntaxException::class)
    fun remove(context: CommandContext<CommandSourceStack>): Int {
        val player = context.getPlayerObject() ?: return 0
        val homeName = context.getArgument("home", String::class.java)
        when (player.removeHome(homeName)) {
            PlayerObject.HomeRemoveResult.NO_HOME -> {
                context.source.sendSystemMessage(Component.translatable("text.homabric.no_home").withStyle(ChatFormatting.RED))
            }
            PlayerObject.HomeRemoveResult.HOME_REMOVED -> {
                context.source.sendSystemMessage(
                    Component.translatable(
                        "text.homabric.home_removed", Component.literal(homeName).withStyle(
                            ChatFormatting.WHITE
                        )
                    ).withStyle(ChatFormatting.GREEN)
                )
            }
        }
        return 1
    }

    @Throws(CommandSyntaxException::class)
    fun list(context: CommandContext<CommandSourceStack>): Int {
        val player = context.getPlayerObject() ?: return 0
        val gui: SimpleGui = player.getHomesGUI(context.source)
        gui.open()
        return 1
    }

    @Throws(CommandSyntaxException::class)
    fun allow(context: CommandContext<CommandSourceStack>): Int {
        val player = context.getPlayerObject() ?: return 0
        val homeName = context.getArgument("home", String::class.java)
        val allowedPlayer = EntityArgument.getPlayer(context, "player")
        when (player.allowHome(homeName, allowedPlayer)) {
            PlayerObject.HomeAllowResult.NO_PLAYER -> {
                context.source.sendSystemMessage(
                    Component.translatable("text.homabric.no_player_exists").withStyle(ChatFormatting.RED)
                )
            }
            PlayerObject.HomeAllowResult.NO_SELF_ALLOW -> {
                context.source.sendSystemMessage(
                    Component.translatable("text.homabric.allow_self").withStyle(ChatFormatting.BLUE)
                )
            }
            PlayerObject.HomeAllowResult.NO_HOME -> {
                context.source.sendSystemMessage(Component.translatable("text.homabric.no_home").withStyle(ChatFormatting.RED))
            }
            PlayerObject.HomeAllowResult.ALREADY_ALLOWED -> {
                context.source.sendSystemMessage(
                    Component.translatable("text.homabric.already_allowed").withStyle(ChatFormatting.RED)
                )
            }
            PlayerObject.HomeAllowResult.HOME_ALLOWED -> {
                context.source.sendSystemMessage(
                    Component.translatable(
                        "text.homabric.allowed", Component.literal(homeName).withStyle(
                            ChatFormatting.WHITE
                        ), Component.literal(allowedPlayer.gameProfile.name).withStyle(ChatFormatting.AQUA)
                    ).withStyle(
                        ChatFormatting.GREEN
                    )
                )
            }
        }
        return 1
    }

    @Throws(CommandSyntaxException::class)
    fun disallow(context: CommandContext<CommandSourceStack>): Int {
        val owner = context.getPlayerObject() ?: return 0
        val homeName = context.getArgument("home", String::class.java)
        val player = context.getArgument("player", String::class.java)
        when (owner.disallowHome(homeName, player)) {
            PlayerObject.HomeDisallowResult.NO_PLAYER -> {
                context.source.sendSystemMessage(
                    Component.translatable("text.homabric.no_player_exists").withStyle(ChatFormatting.RED)
                )
            }
            PlayerObject.HomeDisallowResult.NO_HOME -> {
                context.source.sendSystemMessage(Component.translatable("text.homabric.no_home").withStyle(ChatFormatting.RED))
            }
            PlayerObject.HomeDisallowResult.NOT_ALLOWED -> {
                context.source.sendSystemMessage(
                    Component.translatable("text.homabric.no_player_disallow").withStyle(ChatFormatting.RED)
                )
            }
            PlayerObject.HomeDisallowResult.HOME_DISALLOWED -> {
                context.source.sendSystemMessage(
                    Component.translatable(
                        "text.homabric.disallowed", Component.literal(homeName).withStyle(
                            ChatFormatting.WHITE
                        ), Component.literal(player).withStyle(ChatFormatting.AQUA)
                    ).withStyle(ChatFormatting.GREEN)
                )
            }
        }
        return 1
    }
}